adding doc on working with security advisories (#1263)
chainguard-dev/edu: content/chainguard/chainguard-images/security-advisories/index.md

---
title: "How to Use Chainguard Security Advisories"
linktitle: "Security Advisories"
type: "article"
description: "Article outlining how one can explore and use the Security Advisories found on the Chainguard Image Directory."
date: 2023-12-27T11:07:52+02:00
lastmod: 2023-12-27T11:07:52+02:00
draft: false
tags: ["Overview", "Product", "Chainguard Images", "CVE"]
images: []
menu:
  docs:
    parent: "chainguard-images"
weight: 740
toc: true
---

> Say you use an image scanning tool like Grype or Docker Scout to determine what, if any, vulnerabilities are present within a given Chainguard Image. Chainguard Images typically report few or zero CVEs, but some Images — typically our Production Images, which consist of older versions of a given software and are therefore prone to vulnerabilities — may report more.

Reviewer: Not sure we need to call out Production Images specifically. What do you think about this rewording? There also may be an opportunity to be more upfront about what this will be about.... I tried to do this by pulling the second paragraph up a bit

```suggestion
When you're checking to determine what vulnerabilities, if any, are present in a Chainguard Image, you may be using an image scanning tool like Grype or Docker Scout. Chainguard Images typically report few or zero CVEs, but when older versions of software are included in a given Image, a scanner may report more. Not all vulnerabilities are equal, and it may not be immediately clear whether a CVE found within a given Chainguard Image is likely to cause security issues in your software.
```

Author: I was struggling to not make Chainguard Images sound like they're riddled with CVEs, your wording is much better here. Thanks so much!

> It may not be immediately clear to users whether a CVE found within a given Chainguard Image is likely to cause security issues in their software. After all, scan results can include [false positives](/chainguard/chainguard-images/scanners/false-results/), or true positives may be removed from later versions of the Image. To help demystify the nature of CVEs within Chainguard Images, we've created a self-service [Security Advisories page](https://images.chainguard.dev/security) that lists every security advisory published for Chainguard Images. Having this information available allows you to view whether Chainguard is aware of a specific vulnerability reported to exist within a Chainguard Image and whether we've mitigated or are planning to mitigate the CVE.

Reviewer: I think you can point to the links at the end of the article rather than front-loading them, so we can get to the Advisories content faster

```suggestion
To help demystify the nature of CVEs within Chainguard Images, we've created a self-service [Security Advisories page](https://images.chainguard.dev/security) that lists every security advisory published for Chainguard Images. Having this information available allows you to view whether Chainguard is aware of a specific vulnerability reported to exist within a Chainguard Image and whether we've mitigated or are planning to mitigate the CVE.
```

> It may not be immediately clear to users whether a CVE found within a given Chainguard Image is likely to cause security issues in their software. After all, scan results can include [false positives](/chainguard/chainguard-images/scanners/false-results/), or true positives may be removed from later versions of the Image. To help demystify the nature of CVEs within Chainguard Images, we've created a self-service [Security Advisories page](https://images.chainguard.dev/security) that lists every security advisory published for Chainguard Images. Having this information available allows you to view whether Chainguard is aware of a specific vulnerability reported to exist within a Chainguard Image and whether we've mitigated or are planning to mitigate the CVE.

Reviewer: I think we can frontload more that scanners may misreport vulnerabilities in Chainguard Images, and that you should check the Security Advisories page for any that you want to learn more about, as many of them have been fixed.

> This guide outlines how you can use Chainguard's Security Advisories to learn more about the status of a CVE within a given package. It will walk through  practical example of discovering a vulnerability in a Production Chainguard Image, searching for Security Advisories associated with this vulnerability, and then comparing the original Image with a later version.

Reviewer: ```suggestion
This guide outlines how you can use Chainguard's Security Advisories to learn more about the status of a CVE within a given package. It will walk through a practical example of discovering a vulnerability in a Production Chainguard Image, searching for Security Advisories associated with this vulnerability, and then comparing the original Image with a later version.
```

## Prerequisites

You don't need any special access or software to explore Chainguard's Security Advisories. However, this guide includes a few examples that use specific software tools in order to outline a practical example of how one might navigate and use the Security Advisories.

To follow along with these examples, you'll need the following tools installed.

> * [Docker Scout](https://docs.docker.com/scout/) — A component of the Docker container platform, Docker Scout allows you to scan container images to identify potential vulnerabilities. To use Docker Scout, you will need to [install Docker](https://docs.docker.com/get-docker/) on your chosen platform.

Reviewer 2: Can we change this to something like:

"A security scanner such as Trivy, Grype or Docker Scout. In the following examples we use Docker Scout...."

It's just that scout is commercial so some people won't want to use it.

* [`chainctl`](/chainguard/chainctl/) — Chainguard's command-line interface tool. To install `chainctl`, follow our [installation guide](/chainguard/administration/how-to-install-chainctl/). 
* [`jq`](https://jqlang.github.io/jq/) — `jq` is a command-line JSON processor that allows you to filter and manipulate streaming JSON data. Although it isn’t strictly necessary for the purposes of this guide, this tutorial includes commands that use `jq` to filter command output that would otherwise be difficult to read. You can install `jq` by following the instructions on [the project’s Download jq page](https://jqlang.github.io/jq/download/).

## So you've encountered a CVE in a Chainguard Image

Say you use a vulnerability scanner like Grype or Docker Scout to inspect a certain Chainguard Image. This example uses Docker Scout to scan a Chainguard Production Image, specifically `go:1.21.2`.

```sh
docker scout cves cgr.dev/chainguard-private/go:1.21.2
```

> Because this is the digest for an older version of Chainguard's Go Image, this command's output will show a number of vulnerabilities that have been found to exist within this specific version of the Image. 

Reviewer 2: "this is the digest" -- you didn't specify a digest. I think it would probably best if you did; the tag 1.21.2 will be withdrawn at some point, but the digest would remain available.

Reviewer 2: I'd also add something like "all of these vulnerabilities have been addressed in newer versions of the image"

Author: good points all around!

```
. . .

20 vulnerabilities found in 6 packages
  UNSPECIFIED  8  
  LOW      	1  
  MEDIUM   	3  
  HIGH     	6  
  CRITICAL 	2  
```

This output shows that this particular image has 20 vulnerabilities. The Docker Scout output also lists each of the packages affected by CVEs as well as the specific vulnerabilities it found for each.

> Scrolling to the listing immediately before the final vulnerability count, we'll find that in the case of this example it references the package `nghttp2`. 

Reviewer: ```suggestion
Scrolling to the listing immediately before the final vulnerability count, we'll find that in the case of this example, the package `nghttp2` is referenced. 
```

## Searching the Security Advisories

After finding a vulnerability in a Chainguard Image, you can navigate to [Chainguard's **Security Advisories** page](https://images.chainguard.dev/security). This is a helpful resource you can use to determine the status for any CVE found within a Chainguard Image.

The Security Advisories page is self-service, allowing you to check whether Chainguard is aware of a specific vulnerability and whether it has been mitigated in a certain package version. You can search the Security Advisories page by entering any CVE identifier to find what packages are affected by that CVE. You can also enter the names of individual packages to find what CVEs have been reported within them.

Enter the CVE identifier you copied previously into the search box at the top of the page. This will immediately filter the list of security advisories to only show packages where that CVE has been reported. It will also show the **Status** of each.

![Screenshot showing the Chainguard Security Advisories page. The list of advisories has been filtered to only include those with a CVE ID of "CVE-2023-44487".](sec-advisories-1.png)

> If you click on any row in the filtered list, it will take you to the CVE's specific page. There, you'll find a list of every package where this CVE has been reported.

Reviewer 2: (offtopic) This is correct, but it's a little confusing, so I wonder if it will change in the future...

Author: Yeah I was struggling with describing things in this section. The joys of documenting UI flows. Definitely something to keep tabs on.

![Screenshot showing the Security Advisory page for CVE-2023-44487 and a partial list of its affected packages.](sec-advisories-2.png)

> As with the Security Advisories Landing page, you can filter by package name here on the CVE's landing page as well. Enter the name of the package we highlighted previously (`nghttp2`) and the index will immediately filter out any packages that do not mention the string in their metadata.

Reviewer: ```suggestion
As with the Security Advisories landing page, you can filter by package name here on the CVE's landing page as well. Enter the name of the package we highlighted previously (`nghttp2`) and the index will immediately filter out any packages that do not mention that string in their metadata.
```

![Screenshot showing a filtered list of packages affected by CVE-2023-44487, this time only showing packages that reference "nghttp2". The row for the nghttp2 package itself shows a "Status" of "Fixed", and a "Fixed Version" of "1.57.0-r0".](sec-advisories-3.png)

As the previous screenshot highlights, for CVE-2023-44487, the `nghttp2` package's **Status** is marked as **Fixed** in version `1.57.0-r0` as of October 11, 2023.

## Comparing Images

Chainguard's Security Advisories have told us that the CVE-2023-44487 was fixed and removed from `nghttp2` with a more recent version than the one available in Chainguard's `go:1.21.2` Image. However, we don't have to take that report at face value; we can inspect a later version of the same Image and compare it with version `1.21.2` to determine whether the vulnerability is still present in the later version.

> If you inspect a later version of the Image with Docker Scout, you'll find that this time it reports no CVEs. This example inspects version `1.21.5` of the Image.

Reviewer 2: Probably worth mentioning that at the time of writing 1.21.5 was current, but it might not be now, and may have since accumulated CVEs.

Author: great call

────────────────────┼─────────────────────────────────────────
  Target        	│  cgr.dev/chainguard-private/go:1.21.5   
	digest      	│  65008b35ef40                      	 
	platform    	│ linux/amd64                        	 
	vulnerabilities │	0C 	0H 	0M 	0L         	 
	size        	│ 232 MB                             	 
	packages    	│ 132                                	 

## Packages and Vulnerabilities

  No vulnerable packages detected
```

You can go a step further by comparing these two images directly with the `chainctl images diff` command, as in this example.

```sh
chainctl images diff \
cgr.dev/chainguard-private/go:1.21.2 \
cgr.dev/chainguard-private/go:1.21.5 | jq .
```

This example will return a lot of output, as there are significant differences from version `1.21.2` to `1.21.5` of the Go Image. If you scroll down to the `vulnerabilities` section of this output, you'll find a list of vulnerabilities that are present in version `1.21.2` but have been removed by version `1.21.5`.

```
  "vulnerabilities": {

. . .
  	
{
    	"id": "CVE-2023-44487",
    	"reference": "chainguard:distro:chainguard:rolling",
    	"severity": "High"
  	},

	. . .
```

> As this output indicates, `CVE0223-44487` is no longer present in later versions of the Go Chainguard Image. If you were using version `1.21.2`, you might seriously consider upgrading to a later version.

Reviewer 2: might -> should ?
